### PR TITLE
[20849] Reorganize and speed up test suite

### PR DESCRIPTION
**Pull request by basiszwo**
https://community.openproject.org/work_packages/20849

I did some reorganizing and broke up the test matrix. We are down to 20mins 
runtime. In case we can get more parallel jobs to run (currently 5) we can
decrease even more. 
#### What has been done?
- do not ignore schema.rb and make use of it
- generate coverage report only on demand
- remove object daddy and fully switch to factory girl
- largely split test matrix
- only test against postgresql (myssql should not be checked on each run)
- fix rspec require statements
- extract rspec configs to separate files
- update rspec config
- run most of the specs with capybara-webkit except selenium dependent ones
- reorganize rspec files
  - combine api specs
  - divide rspec feature specs in modules (1st step towards modules and required for build matrix)
- reorganize cucumber features: divide into modules (mainly for build matrix)
- add license to some files
- add parallel_test gem
- require features in cucumber config to be able to run `cucumber features/foo/bar.feature`
- do not create screenshots for failing specs in CI
#### Next steps
- Test parallel_test on travis and see if this decreases test time even more.
- run webpack only once
- compile asset before
- identify slowest tests and rewrite them

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://basiszwo-20849-test-suite.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://basiszwo-20849-test-suite.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by myabc** [comment]
> Remove legacy specs as they are not being run anyway

This is not true. See Travis.

File: .travis.yml [hunk 1]
```diff
@@ -79,9 +100,7 @@ script:
 
 notifications:
   email: false
-  webhooks:
-    urls:
-      - "https://webhooks.gitter.im/e/435ab5d5c6944305da2f"
+  slack: openprojectservices:VF0xFQjYrBWDMFveAtUtxJDL
```
Comment (Azure7111) [inline .travis.yml]: There is a slack now for OpenProejct?

**Comment by ulferts** [comment]
I am pretty sure that e.g. https://github.com/opf/openproject/blob/dev/spec/legacy/functional/user_mailer_spec.rb covers code we haven't covered otherwise. I'd therefore like to process a tad more selectively when it comes to removing specs. 

We could for example check whether we at least have a corresponding file in our non legacy test suite before removing a file from the legacy suite.

**Comment by ulferts** [comment]
This is a pretty big PR bringing Github to it's limit. I am therefore refraining from commenting on individual changes and rather write down what I noticed in one comment. It might make sense to break parts of this PR out into smaller ones and squash other changes to make it reviewable.

# I like
- Splitting up the tests into smaller chunks should make it easier to execute them separately and also it should give some more structure. Maybe we'll one day be able to break chunks of the application out of here completely. That would be the most efficient speedup we can get.
- Structuring the tests by functional groups (e.g. work_package, wiki)  
- Cleaning up the codebase (structure and removal of old dependencies).

# I wish
- That the breaking up into functional groups would also be applied to the non feature specs. We now still have a huge glob of model/controller/view specs. They'll need to be broken up into smaller pieces still. And IMO it should also be by feature. Is a big pile of work though, so I already appreciate what you have done so far.
- We'd also enforce breaking up by the spec/test_helper that is available within each block. We could have a test helper for tests that only supports unit tests in their strict definition (No DB/FS access), one for unit tests requiring DB access, one for integrational/functional (in the sense of testing at least two classes working together) and one for E2E tests. Those would then only load the parts of the application that is required for running the particular type of test leading to fast ramp up time. Additionally, we would force developers into a schema within each test type.  
- That the `@wip` tags added where removed again. 

# I fear
- The overhead in setup time when breaking up the tests in smaller chunks like this especially on the internal CI. 
- That it will take additional effort to make it work on the internal CI with the configurations (OP core + plugins) we maintain. We'll have to handle the different schema.rb per configuration.
- That running the tests only on postgresql will lead to failures being discovered by our internal CI after they have been merged in.
- That the removal of snapshots on the CI will make it harder to find flickering specs. That feature was introduced particularly to understand such failures. If they are to be disabled, they should be disabled when not running on a CI. I don't care for them locally as I can debug. 

# Additional notes
- There are references to parallel_test in there although it is not used yet. Do you plan on employing it within this PR? 
- Before merging, this needs to be executed on the internal CI for every configuration we have at least once.

**Comment by NobodysNightmare** [comment]
@basiszwo wrote:

> only test against postgresql (myssql should not be checked on each run)

@ulferts wrote:

> I fear [...] That running the tests only on postgresql will lead to failures being discovered by our internal CI after they have been merged in.

I agree with the fear of ulferts. Code that is not fully DB agnostic is very well possible. Just today there was an example of code only working on PSQL and not for MySQL, I myself have written code taht failed for PSQL, but worked on MySQL. I see that those double runs are a bit of a pain, but at the same time I feel they are neccessary.

When not asking **if** we want to disable one of two adapters for travis runs, I would be asking **which** adapter to disable. Is there a rationale to choose PSQL over MySQL? Or is that a decision based on a gut feeling?

File: Gemfile.lock [hunk 2]
```diff
@@ -590,6 +592,3 @@ DEPENDENCIES
   warden (~> 1.2)
   warden-basic_auth (~> 0.2.1)
   will_paginate (~> 3.0)
-
-BUNDLED WITH
-   1.10.6
```
Comment (NobodysNightmare) [inline Gemfile.lock:595]: @basiszwo Please update your bundler version. That this line is removed is an annoyance that the bundler team describes here:

http://bundler.io/blog/2015/06/24/version-1-10-released.html#bundled-with

**tl;dr:** New bundlers will add this line, old bundlers will remove it: **Git edit wars!** Update to a recent bundler: Worst case of one change per bundler version.

edit: for reference: the offending commit is https://github.com/basiszwo/openproject/commit/30a559da75e81d50d5af59c9bd51cf5ac8c8eee4

File: lib/tasks/ci.rake [hunk 3]
```diff
@@ -81,10 +81,11 @@ namespace :ci do
...
       # depends on an existing databse.
       Rake::Task['environment'].invoke
-      Rake::Task['db:load_config'].execute
-
-      Rake::Task['db:migrate'].invoke
-      Rake::Task['db:schema:dump'].invoke
+      # Rake::Task['db:load_config'].execute
+      #
+      # Rake::Task['db:migrate'].invoke
+      # Rake::Task['db:schema:dump'].invoke
```
Comment (NobodysNightmare) [inline lib/tasks/ci.rake]: Can we safely delete those commented lines?

**Comment by NobodysNightmare** [comment]
Something I noticed for the first time:

> [DEPRECATION] Capybara::Webkit::Driver#browser is deprecated.

Is this new to your PR? I'd say yes. Found here: https://travis-ci.org/opf/openproject/jobs/75937099

(There are other deprecation warnings in our tests... not saying those are the first deprecation warnings at all)

**Comment by basiszwo** [comment]
@NobodysNightmare as for the deprecation warning see https://github.com/thoughtbot/capybara-webkit/pull/778 I hope to get rid of the warnings soon.

**Comment by basiszwo** [comment]
closing this pull request for further testing that should not pollute the official build pipeline.
